GQSS-4899: Working network stats on VMware.

ProjectMoon opened the pull request:
This commit changes the logic in the VmwareResource to properly pull
down network TX/RX statistics. Previously the code attempted to query
all available performance metrics, and if it could find them, it would
add up the network info counters. The code has been changed to
directly query for the network read/write performance counters using
the QueryPerf API, using the approximately one minute intervals as the
interval to query for.

The method to retrieve VM stats has also been significantly
refactored, with each of the various steps in gathering the stats
separated into individual methods.

ProjectMoon commented:
Tests are not yet ready. Some unit tests can be delivered though, by mocking the VMware API and making sure the logic returns a sane number in a few different scenarios.

HrWiggles commented on plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -4859,103 +4877,215 @@ private VirtualMachineGuestOsIdentifier translateGuestOsIdentifier(String cpuArc
+        //We need to calculate the number of 20-second samples in the interval. Since we are using the 20 second
+        //intervals (see queryNetworkPerfs), and poll once per minute, this number will be around 3.
+        double numSamples = Seconds.secondsBetween(jodaThen, jodaNow).getSeconds();
+        numSamples = numSamples * (numSamples / 20.0);
```
I don't quite understand your calculations...
You start by getting the number of seconds between `now` and `then` which, if polling is done once per minute, should be ~60.
You then divide those seconds by 20.0 (which, as you state in the above comment, will be around 3), but then multiply the result against the original number of seconds, which should make value of `numSamples` be ~180.  What significance would the value 180 have?  Shouldn't a variable named `numSamples` be set as the number of samples? (which one would think would be 3 in this case)
Furthermore, in the calculations below, you loop through a series of integers, adding them all up, and then divide that sum by the number of integers in the series (at least, that's what I believe you're doing - in which case, there is a bug in the code as you divide by `metric.getValue().size()` instead of `intSeries.getValue().size()`).  What are those series of integers?  Are they sizes corresponding to each 20 second interval?  If so, shouldn't they be divided by 20.0 in the end? (instead of `numSamples`)  Also, assuming that the series of integers are sizes corresponding to each 20 second interval, would there be any need for the `numSamples` variable?  Wouldn't `intSeries.getValue().size()` tell you the number of samples?

One last thing: can either `numSamples` or `metric.getValue().size()` be zero? (or `intSeries.getValue().size()` if I'm right about there being a bug in the code)  If so, there need to be guards against dividing by these values (i.e. avoid divide-by-zero)

HrWiggles commented on plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -4859,103 +4877,215 @@ private VirtualMachineGuestOsIdentifier translateGuestOsIdentifier(String cpuArc
+                        average += val;
+                    }
+
+                    average = average / metric.getValue().size();
```
Why divide by `metric.getValue().size()`?  Shouldn't it be dividing by `intSeries.getValue().size()` instead?

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
+
+    @Test
+    public void testVmNetworkStatsNoMetrics() throws Exception {
+        //TODO modify mockMetric so that it can return a PerfCounterInfo but not PerfEntityMetric
```
Was this comment supposed to be committed?
Seems like this is no longer an issue since you use `mockPerfCounterInfo` and `mockMainStuffWithoutMetrics`.

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
+        ManagedObjectReference vmMor = mock(ManagedObjectReference.class);
+
+
+
```
Inconsistent whitespace: two extra empty lines compared to the other two tests.

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
+
+        Pair<PerfCounterInfo, PerfEntityMetric> txPair = mockMetric("net", "transmitted", 148, 10);
+        Pair<PerfCounterInfo, PerfEntityMetric> rxPair = mockMetric("net", "received", 148, 10);
+        List<XMLGregorianCalendar> xmlCals  = mockMainStuff(Arrays.asList(txPair, rxPair), then, now);
```
Extraneous space character after `xmlCals`
This is the case for all three tests.

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
+        Pair<Double, Double> networkStats = _resource.calculateNetworkReadWrite(vmMor, then, now);
+        Assert.assertNotNull(networkStats);
+        Assert.assertTrue(networkStats.first() > 0);
+        Assert.assertTrue(networkStats.second() > 0);
```
Why not check for the expected values?  Only checking that the values are greater than zero does not really fill one with confidence that the calculations are correct.
Any reason why this couldn't simply do `Assert.assertEquals(5500.0, networkStats.first(), 0.0);`? (and, of course, the same for `networkStats.second()`)

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
 
+    @Test
+    public void testVmNetworkStats() throws Exception {
+        //We need to mock an interval of one minute somehow...
```
This comment (and the identical one in `testVmNetworkStatsZero`) feels unnecessary, even misleading (since there doesn't appear to be any mocking of anything so that `now` would be considered 60 seconds later than `then`).
Maybe just drop these two comments?

HrWiggles commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -98,4 +231,57 @@ public void testScaleVMF1() throws Exception {
+
+
+        PerfCounterInfo txCounter = mockPerfCounterInfo("net", "transmitted", 148);
+        PerfCounterInfo rxCounter = mockPerfCounterInfo("net", "received", 149);
```
Are these IDs (`148` and `149`) significant?
Were the IDs supposed to be different here? (seeing as ID `148` is used everywhere except for `rxCounter` here)
If ID `149` was a typo (and the ID should always be `148`), then maybe just skip it as a parameter?
Either way, would be nice if these IDs would be made into named constants, if only to improve readability.

HrWiggles commented on plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -4859,105 +4879,211 @@ private VirtualMachineGuestOsIdentifier translateGuestOsIdentifier(String cpuArc
-                        continue;
-                    }
+                        if (intSeries.getValue().size() > 0) {
+                            average = average / intSeries.getValue().size();
```
Could move initialization of variable `average` and the loop over `intSeries.getValue()` into the `if`-block:
```
for (PerfMetricSeries series : metric.getValue()) {
    if (series instanceof PerfMetricIntSeries) {
        PerfMetricIntSeries intSeries = (PerfMetricIntSeries)series;
        if (intSeries.getValue().size() > 0) {
            double average = 0.0;
            for (Long val : intSeries.getValue()) {
                average += val;
            }

            average = average / intSeries.getValue().size();
```
Also, maybe use a pre-existing implementation for calculating the average, e.g. `com.google.common.math.DoubleMath`
That way, this code:
```
            double average = 0.0;
            for (Long val : intSeries.getValue()) {
                average += val;
            }

            average = average / intSeries.getValue().size();
```
could simply be replace with
```
            double average = DoubleMath.mean(intSeries.getValue());
```

ProjectMoon commented:
Addressed the comments. The ID numbers are from one of our VMware test systems. Not sure if they change between VMware setups or not, but they do need a number.
